feat: Add EdgeX-CLI-V2 ADR

siggiskulason opened the pull request:
add edgex-cli V2 ADR

This PR adds a new  ADR with a technical plan for creation of a 2.0 version of edgex-cli which supports the new V2 REST APIs.

Signed-off-by: Siggi Skulason <siggi.skulason@canonical.com>

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] Changes have been rendered and validated locally using mkdocs-material (see edgex-docs README)
- [X] Docs have been added / updated (for bug fixes / features)

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/edgexfoundry/edgex-docs/blob/master/.github/Contributing.md.

## Issue Number:

## What is the new behavior?
ADR with a technical plan for edgex-cli 2.0.

## Does this PR introduce a breaking change?
<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

- [ ] Yes
- [X] No

## Specific Instructions
Are there any specific instructions or things that should be known prior to reviewing?


## Other information

tonyespy commented on docs_src/design/adr/core/0009-V2-EdgeX-CLI.md:
```diff
@@ -0,0 +1,66 @@
+
+5. Implement support for use of the client via the API Gateway, including being able to connect to a remote EdgeX instance.
+
+6. Implement all required V2 endpoints as per current [Issue](https://github.com/edgexfoundry/edgex-cli/issues) list.
```
I'd suggest just including the services and associated commands to implement inline vs. referencing the issues.  Ex.

Core Command
  - `command  get | put | list`  (**note**,  we might want replace `get/put` with `read/write` to be more consistent w/other commands)

Core Data
  - `event add | count | list | rm | scrub`**
  - `reading count | list`

Two comments here, I think `scrub` directly deletes events from the database, and I'm not sure this will work with Redis being secured by default. I would mention this and see what others think. The second is that I think `rm` should be re-named to `delete` or `remove`.

Metadata

 - `device add | adminstate | list | operstate | rm | update`
 - `profile` (should this be deviceprofile for consistency?) `add | list | rm | update`
 - `deviceservice add | list | rm | update`
 - `watcher` (same question re: provisionwatcher or provwatcher?) `add | list | rm` (why no update?)
 
.
.
.

tonyespy commented on docs_src/design/adr/core/0009-V2-EdgeX-CLI.md:
```diff
@@ -0,0 +1,66 @@
+    - `edgex data ping`
+    - `edgex metadata version`
+    - `edgex metadata metrics`
+
```
Let's discuss. If we do make this change, I'd suggest mentioning that the <service> could be an abbreviation or full name. Ex.

```
$ edgex (d)ata ping
$ edgex (m)etadata ping
```

I suppose tab-completion gives you this in a shell, but this would simplify scripting.

That said, we should prepare for some pushback on this, and if so, at least offer an alternative flags-based solution for the common (e.g. `metrics`, `ping`, ...) endpoints.

tonyespy reviewed (commented):
Looks good, but please see comments inline.  Also three high-level comments:

  - You need to add the ADR to the [TOC](https://github.com/edgexfoundry/edgex-docs/blob/master/docs_src/design/TOC.md) and number your new ADR accordingly (i.e. looks like the next free number is 19)
  - Please also add an actual PR description and feel free to remove the non-applicable bits from the PR template
  - I'd include _V2_ in your commit message

siggiskulason commented on docs_src/design/adr/core/0009-V2-EdgeX-CLI.md:
```diff
@@ -0,0 +1,66 @@
+
+## Context
+This ADR presents a technical plan for creation of a 2.0 version of the edgex-cli which supports the new V2 REST APIs developed as part of the Ireland release of EdgeX.
+Existing Behavior
```
done

siggiskulason commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,75 @@
+| ----------- | ----------- |
+| `-d`, `--debug`      | show additional output for debugging purposes (e.g. REST URL, request JSON, …). This command-line arg will replace -v, --verbose and will no longer trigger output of the response JSON (see -j, --json).       |
+| `-j`, `--json`   | output the raw JSON response returned by the EdgeX REST API and *nothing* else. This output mode is used for script-based usage of the client.    |
+| `--version`   | output the version of the client and if available, the version of EdgeX installed on the system.    |
```
It uses the core data version - ADR updated.

tonyespy commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,125 @@
+    - **`version`**
+    - **`ping`**
+    - **`metrics`**
+    - **`status`** 
```
nit: I would just add a dedicated section for the common endpoints vs. repeating them across all services...

jpwhitemn commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+
+3. Take full advantage of the features of the underlying command-line library, [Cobra](https://github.com/spf13/cobra), such as tab-completion of commands.
+
+4. Change the client binary name from `edgex-cli` to `edgex`. The reason is that the "-cli" is superfluous. There is no `edgex` tool and best practice is to choose a simple, memorable word. CLI tool names do not generally include `"-cli"` in the name. (Examples: `go`,`snap`,`kubectl`,`git`,`docker` etc)
```
I am in the camp of removing '-cli'.  Other frameworks/tools would not include "cli" unless there were multiple types of clients to interact with.  To @lenny-intel - while explicit, what else would it be confused with?

lenny-intel replied:
Would folks run the command `edgex` and expect the services to start?

siggiskulason replied:
@lenny-intel / @jpwhitemn - I am on the fence on this one. Whilst I think the "-cli" might be superfluous, I see the point about `edgex-cli` being explicit and removing it might give the mistaken impression that running `edgex` by itself would be a way to start up the core services. Also, as we have `edgex-ui` it makes sense to have a corresponding `edgex-cli`. I have therefore removed this suggestion and suggest we stick with `edgex-cli` and don't change it for now.

lenny-intel replied:
Works for me :-)

jpwhitemn commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+
+5. Add support for the four common endpoints - `/ping`, `/metrics`, `/version`, `/status`
+
+6. Change the syntax of the CLI to require a positional service arg which precedes the command, thus allowing overlap of command names across services. The service name can either be an abbreviation or the full name of the service . Examples:
+
+    - `edgex data ping`
+    - `edgex d ping`
+    - `edgex metadata version`
+    - `edgex m version`
+    - `edgex metadata device list`
+    - `edgex command device list`
+
+    An alternative solution is to use a positional argument only for the common endpoints but specify it after the command or to have a regular arg like `-s`/`--service` `<name>`).
```
This was discussed before with the VMWare team.  At the time, it was decided that the service was implicit in many commands because the request was only germane to a service.  
Would it be possible to do a little of each?  For example, if I did a request for events, that would not require the service name (either preceeding or post with a -s) because the service would be known to be associated to core data.  But if I asked for a device list, then the CLI would come back with an error indicating multiple services have that command and a more explicit service name is required.  Service name could always be provided from the beginning for extra clarity.  As to whether it is positional or after a -s or --service, I am for the latter.  Position arguments always confuse me when other parameters are also positional.

siggiskulason replied:
Thanks for the comments. Yes, we discussed this in the Core WG meeting a week ago and I would tend to now agree what you both suggest, i.e. having a -s/--service for the cases where there is more than one possible service that provides the endpoint, but otherwise the CLI would be smart enough to know which one is being used.

siggiskulason replied:
I have implemented this in [this PR](https://github.com/edgexfoundry/edgex-cli/pull/413) as suggested in your comments. For the common endpoints, it's now possible to specify the service to use: `-m/--metadata`, `-c/--command`, `-n/--notification`, `-s/--scheduler` or `--data` (which is the default). I updated this point in the ADR accordingly.

jpwhitemn commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+    - **`ping`**
+    - **`metrics`**
+    - **`status`** 
+
```
I think there are some API endpoints/commands that are higher priority than others.  If we need to limit the functionality to some of the endpoints in this version, I am ok with that.  VMWare did a priority matrix that we can refer to if needed.

siggiskulason replied:
@jpwhitemn - thank you for the comments. I have responded to them inline and updated the ADR.

jpwhitemn commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+
+9.  Currently, some commands default to always displaying GUIDs in objects when they're not really needed. Change this so that by default GUIDs aren't displayed, but add a flag which causes them to be displayed. 
+
+10. **scrub** may not work with Redis being secured by default. That might also apply to the top-level `db` command (used to wipe the entire db). If so, then the `db` command will be removed.
```
Are we assuming this V2 CLI is going to work with security on?  Existing CLI does not.  I am ok if first release of V2 only works with security off.  For the most part, this is a developers tool.  Or make this a stretch goal.

From a developers standpoint, the scrub command is important.  Where a command is not capable to run because of security, we should still offer it for non-secure mode.

siggiskulason replied:
Yes - as a stretch goal. I updated this to say "*scrub** may not work with Redis being secured by default. That might also apply to the top-level `db` command (used to wipe the entire db). If so, then the commands will be disabled in secure mode, but permitted in non-secure mode. "

jpwhitemn commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+
+12. *(Stretch)* Add registry support instead of requiring static configuration file. Alternatively, hard-code the default ports of the reference services and provide a `set-hostnames` command to select the default values, taking `localhost`, `docker-network`, or `api-gateway` as options. In that case, also modify each command to support `-p`/`--port` and `-h`/`--host arguments` for custom configu ...
+
+13. *(Stretch)* Implement support for use of the client via the API Gateway, including being able to connect to a remote EdgeX instance. This might require updates in go-mod-core-contracts.
```
+1 as stretch.  Push this to next release if it becomes big lift.

jpwhitemn reviewed (commented):
@siggiskulason - I really like the proposed changes.  Argument flags for -j and -d make a lot of sense.

I think the discussion on how to best provide the service name when it is needed requires a group discussion/decision.  I have a preference, but would defer to the developer community consensus.

Really good work - thank you.!

siggiskulason commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,113 @@
+
+11. *(Stretch)* implement a `-o`/`--output` argument which could be used to customize the pretty-printed objects (i.e. non-JSON).
+
+12. *(Stretch)* Add registry support instead of requiring static configuration file. Alternatively, hard-code the default ports of the reference services and provide a `set-hostnames` command to select the default values, taking `localhost`, `docker-network`, or `api-gateway` as options. In that case, also modify each command to support `-p`/`--port` and `-h`/`--host arguments` for custom configu ...
```
Agree - with a static config we also no longer need a configuration file at all, which means that point 8 about using Viper vs BurntSushi can also be removed.

siggiskulason replied:
ADR updated accordingly in section 9.

siggiskulason commented:
I have rebased the ADR and updated it in response to the review comments (thank you, @lenny-intel, @tonyespy  and @jpwhitemn ). Would you be able to re-review, please? I have also submitted a [PR](https://github.com/edgexfoundry/edgex-cli/pull/413)  implementing the first feature set as per this ADR - with more to come.

lenny-intel commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,104 @@
+6. Implement all required V2 endpoints for core services
+
+    **Core Command** (`edgex-cli (c)ommand`)
+    - **`command`** `get | put | list` *(we might want to replace get/put with read/write to be more consistent w/other commands)*
```
These should be `read/write` to be consistent with the Device Command terminology.

siggiskulason replied:
updated

lenny-intel commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,104 @@
+    - **`command`** `get | put | list` *(we might want to replace get/put with read/write to be more consistent w/other commands)*
+
+    **Core Data** (`edgex-cli (d)ata`)
+    - **`event`** `add | count | list | rm | scrub**`
+    - **`reading`** `count | list`
```
Will `list` command allow options like specifying device name?

siggiskulason replied:
yes, 
` event list` would use /event/all
`event list --device {name}` would use /event/device/name/{name} to return the events sourced from the specified device 
' 
etc. for other arguments. I'll add this to the ADR.

lenny-intel commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,104 @@
+| ----------- | ----------- |
+| `-d`, `--debug`      | show additional output for debugging purposes (e.g. REST URL, request JSON, …). This command-line arg will replace -v, --verbose and will no longer trigger output of the response JSON (see -j, --json).       |
+| `-j`, `--json`   | output the raw JSON response returned by the EdgeX REST API and *nothing* else. This output mode is used for script-based usage of the client.    |
+| `--version`   | output the version of the client and if available, the version of EdgeX installed on the system (using the version of the core data service)   |
```
```suggestion
| `--version`   | output the version of the client and if available, the version of EdgeX installed on the system (using the version of the core metadata service)   |
```

lenny-intel replied:
Core Data is technically now optional...

siggiskulason replied:
changed to refer to metadata (which is what the implementation already does).

lenny-intel reviewed (approved):
LGTM

tonyespy commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,106 @@
+
+6. Implement all required V2 endpoints for core services
+
+    **Core Command** (`edgex-cli (c)ommand`)
```
Shouldn't you remove the `edgex-cli (c) command` as this approach is no longer being implemented (in favor of a service-specific argument; e.g.` -c/--command`)?

tonyespy replied:
**Note** - this comment applies to all the other service sub-headers in this section.

siggiskulason replied:
yes - removed the service name from the sub-headers.

tonyespy commented on docs_src/design/adr/core/0019-EdgeX-CLI-V2.md:
```diff
@@ -0,0 +1,106 @@
+    - **`interval`** `add | list | rm | update` 
+
+    **Common endpoints in all services**
+    - **`version`**
```
Aren't these already mentioned in `#4` above? Maybe just delete it, and leave these here in this section?

siggiskulason replied:
deleted.

tonyespy reviewed (changes requested):
Looks much better, but a couple of comments inline still need some attention.

tonyespy reviewed (approved):
LGTM
